### ISSUE TITLE
Allow user to get to cpal errors such as device unplugged.

### ISSUE THREAD

**Issue by tochiu** [RustAudio/rodio#465]
Hi, I'm writing a tui and am using rodio for audio playback.

I'm not sure how to handle `an error occurred on output stream: The requested device is no longer available. For example, it has been unplugged`. This happens when I unplug my earphones which makes sense. I want to retry the default device but I'm not sure how to detect this error because it comes from `eprintln!`. 

Also, I'd like there to be no prints in the first place since it screws up my terminal output. I would think errors are routed through the `log` crate for library users to handle. 

What am I missing on both fronts?

**Comment by est31** [comment]
> I want to retry the default device but I'm not sure how to detect this error because it comes from eprintln!

In theory the error should be returned to the user, and not be routed to `eprintln`.  The error originates from cpal actually, and rodio only transmits it. Can you maybe check your code if it contains eprintln?

**Comment by tochiu** [comment]
I have `eprintln!` only in areas where operations that return the terminal to normal fail which is divorced from audio playback. I doubt it's any of those since the program runs as if no error happened but all audio operations result in no sound which makes sense. All other logging is routed through the `log` crate in my application.

I presumed the error came from this line here: https://docs.rs/rodio/latest/src/rodio/stream.rs.html#198 but I didn't dig any further. 

I use rodio in my project here: https://github.com/tochiu/c8/blob/main/src/run/audio.rs

**Comment by est31** [comment]
@tochiu I see. Yeah there should be a way to customize the callback....

**Comment by photovoltex** [comment]
Problem still seems to occur on `master`. The error was adjusted and is now the following: `error initializing output stream: {err}`.

**Comment by will3942** [comment]
@dvdsk We need this feature, would you happy to receive a PR with this implemented? I was thinking of extending `OutputStreamBuilder` with a `with_error_callback` function.

**Comment by photovoltex** [comment]
I tried something of the like [here](https://github.com/RustAudio/rodio/compare/master...photovoltex:rodio:expose-error-callback). But when I tried to use it, it seems to never notify the receiver. It feels like the thread completly crashes and never achieves to send the information. But maybe my code idea is flawed to begin with.

**Comment by dvdsk** [comment]
> @dvdsk Would you happy to receive a PR with this implemented? 

yes please, though to not waste your time please let us know the API/design your are planning on. We might be able to help refine before you go through the effort of implementing it.

> I was thinking of extending `OutputStreamBuilder` with a `with_error_callback` function.

that seems like a good fit 👍 

> I tried something of the like [here](https://github.com/RustAudio/rodio/compare/master...photovoltex:rodio:expose-error-callback).

technically not a callback, though an mpsc::Sender it probably how most users will want to it. Might be a good idea to feature use of mpsc::Sender in an doc example.

### some observations
- It might be valuable to steer users towards correct usage through examples, maybe filtering in the callback for the error they want?
- Relevant discussion on error handling in rodio (or lack thereof rather): https://github.com/RustAudio/rodio/issues/677.
- We currently have some `eprintln!` / `tracing::error!` calls here and there throughout the code, this would then replace it. The `eprintln`/`error!` callback should probably be the default if no callback is provided? 
- Performance or rather the question of dynamic vs static dispatch (impl vs dyn)

**Comment by will3942** [comment]
> yes please, though to not waste your time please let us know the API/design your are planning on. We might be able to help refine before you go through the effort of implementing it.

The API I was thinking is fairly simple (essentially exposing the cpal type which is potentially a leaky abstraction but no more than the current OutputStream) to keep it flexible for the users, on `OutputStreamBuilder`:

```rust
pub fn with_error_callback<E>(
    mut self,
    callback: E
) -> OutputStreamBuilder
where
    E: FnMut(cpal::StreamError) + Send + 'static
```

Agreed on:
- Adding an example, I think one detailing filtering for the `DeviceNotAvailable` error would suffice
- Default callback would be the current one with `eprintln!` / `tracing::error!`

Let me know if you've got any other ideas or you'd like to discuss/confirm anything else before I implement.

**Comment by photovoltex** [comment]
I initially tried to use a general callback (closure/fn) but that didn't seem to be possible as the trait bounds require a mutable clonable callback... That's why I did fallback to the channel logic. But it's only a poc that didn't even work as the thread seems to die before the channel can send the error anyways.

I think the underlying problem is that in windows you actually create a cpal device for an actual hardware device, where in linux you create a cpal device which represents an audio system like alsa. By that in linux the "device" never actually dies, but in windows the device dies when removed as it represents the actual hardware. But that's all just some broad observation and guesses as I didn't look too deep into the cpal code. 

> Would be interesting to know tho, how close my guess is from the actual logic.

**Comment by will3942** [comment]
> I think the underlying problem is that in windows you actually create a cpal device for an actual hardware device, where in linux you create a cpal device which represents an audio system like alsa. By that in linux the "device" never actually dies, but in windows the device dies when removed as it represents the actual hardware. But that's all just some broad observation and guesses as I didn't look too deep into the cpal code.

That sounds like an issue on the Linux end with hotplugging rather than something we could solve from Rodio? Or at least the PR would be upstream in cpal to react to any ALSA (or other) events and send those through correctly as `DeviceNotAvailable` errors?

**Comment by dvdsk** [comment]
> See also [#677](https://github.com/RustAudio/rodio/issues/677) I already tried to solve this by providing a callback. that can re-create the stream automatically when stream dies, but the code looks rather cumbersome. Currently the stream owns the whole processing chain so if it has to be re-created. Maybe callback can get ownership of the sources on error but then we need a way to detach them from the older callback. If just a call-back is enough then this doable right away.

that would be the best way of handling an unplug. That is pretty difficult though, maybe `cpal` could be changed to allow us to reclaim the source callback? Anyway simply getting the error to the user is already a good improvement to the current situation.
